src/main/kotlin/com/cognifide/gradle/aem/common/pkg/PackageValidator.kt

```diff
@@ -0,0 +1,182 @@
+
+    @Input
+    var verbose = aem.props.boolean("aem.package.validator.verbose") ?: true
+
+    @OutputDirectory
+    var root = File(aem.project.buildDir, "aem/package/validator")
+
+    @Input
+    var planName = aem.props.string("package.validator.opear.plan") ?: "plan.json"
+
+    @get:Internal
+    val planFile get() = File(root, planName)
+
+    @get:Internal
+    val reportFile: File
+        get() = File(root, "oakpal-report.json")
+
+    private var baseProvider: () -> File? = {
+        aem.props.string("package.validator.opear.base")?.let { aem.resolveFile(it) }
+    }
+
+    fun base(value: String) = base { aem.resolveFile(value) }
+
+    fun base(provider: () -> File?) {
+        this.baseProvider = provider
+    }
+
+    @get:InputFile
+    @get:Optional
+    val baseFile: File? get() = baseProvider()
+
+    @get:InputFiles
+    val configDirs get() = _configDirs.filter { it.exists() }
+
+    private var _configDirs = mutableListOf(
+            File(aem.configCommonDir, CONFIG_DIR_PATH),
+            File(aem.configDir, CONFIG_DIR_PATH)
+    )
+
+    fun configDir(dir: File) {
+        _configDirs.add(dir)
+    }
+
+    fun configDirs(dirs: Iterable<File>) {
+        _configDirs = dirs.toMutableList()
+    }
+
+    fun configDirs(vararg dirs: File) = configDirs(dirs.asIterable())
+
+    fun perform(vararg packages: File) = perform(packages.asIterable())
+
+    fun perform(packages: Iterable<File>) {
+        prepareOpearDir()
+        runOakPal(packages)
+    }
+
+    private fun prepareOpearDir() {
+        logger.info("Preparing OakPAL Opear directory '$root'")
+
+        root.deleteRecursively()
+        root.mkdirs()
+
+        baseFile?.let { file ->
+            logger.info("Extracting OakPAL Opear base configuration files from '$file' to directory '$root'")
+            FileOperations.zipUnpackAll(file, root)
+        }
+
+        configDirs.filter { it.exists() }.forEach {configDir ->
+            logger.info("Using project-specific OakPAL Opear configuration files from file '$configDir' to directory '$root'")
+
+            FileUtils.copyDirectory(configDir, root)
+        }
+    }
+
+    private fun runOakPal(packages: Iterable<File>) {
+        val opearFile = OpearFile.fromDirectory(root).getOrElse {
+            throw AemException("OakPAL Opear directory cannot be read properly!")
+        }
+
+        val planUsed = planFile.takeIf { it.exists() }?.toURI()?.toURL() ?: opearFile.defaultPlan
+
+        logger.info("Validating CRX packages(s) '${packages.names}' using OakPAL plan '$planUsed'")
+
+        val scanResult = OakpalPlan.fromJson(planUsed)
+                .map { plan ->
+                    plan.toOakMachineBuilder(DefaultErrorListener(), opearFile.getPlanClassLoader(javaClass.classLoader))
+                            .withNodeStoreSupplier { MemoryNodeStore() }
+                            .build()
+                }
+                .map { oak -> oak.scanPackages(packages.toList()) }
```

Comment:
Please encapsulate in method and give meaningful name

Reply:
tried, but not looking more nice, 3 args required